clash-prelude/src/Clash/Signal/BiSignal.hs

```diff
@@ -183,8 +183,13 @@ data BiSignalIn (ds :: BiSignalDefault) (dom :: Domain) (n :: Nat)
 --
 -- Wraps (multiple) writing signals. The semantics are such that only one of
 -- the signals may write at a single time step.
+#if MIN_VERSION_base(4,15,0)
+data BiSignalOut (ds :: BiSignalDefault) (dom :: Domain) (n :: Nat)
+  = BiSignalOut ![Signal dom (Maybe (BitVector n))]
+#else
 newtype BiSignalOut (ds :: BiSignalDefault) (dom :: Domain) (n :: Nat)
   = BiSignalOut [Signal dom (Maybe (BitVector n))]
+#endif
```

Comment:
What's up with this change? I wouldn't mind making it strict for all versions (changelog necessary though).

Reply:
This is needed because for any newtype `T`. GHC will transform:

```
f :: a -> T
{-# NOINLINE f #-}
```

to:
```
f1 :: a -> RepOfT
{-# NOINLINE f1 #-}

f = f1 |> (co :: a -> RepOfT ~ a w %'Many-> T)
``` 
Which is a `LinearTypes` thing. Note how `f` no longer has the NOINLINE!!! That means we can no longer recognize our primitives. The work-around is to use `data T`. However:

```
data T = MkT a
```
is not the same as:
```
newtype T = MkT a
```
because data-constructors are lazy. Closer in spirit to the newtype is:
```
data T = MkT !a
```
It's still not the same as the newtype, as we do add another machine word to represent the data constructor `MkT`; but the field is no longer a pointer to the thunk `a` but but an actual value `a`. All in all, performance is going to be worse, so it's definitely not something we want for all versions.